### PR TITLE
Feat equilibrator thermodata

### LINKED ISSUE #36: Integration with eQuilibrator
As an user, I would like to use the [eQuilibrator-API](https://gitlab.com/elad.noor/equilibrator-api) as the thermodynamic database. Thus, eQuilibrator would parse the reactions, calculate ΔGr for each reaction and use the data to prepare and convert a `cobra.Model` into a `pyTFA.ThermoModel`. Is there any reason that hampers this integration?

I would be happy to help with the implementation, maybe as an extra dependency. I have used the tutorial in `tutorials/figure_paper.py` to try to reproduce the results just by adding reaction data, but the results differ (at the end of [this notebook](https://github.com/carrascomj/notebooks/blob/master/DD-DeCaF/pyTFA/pytfa_time.ipynb)).

### PR DESCRIPTION
**Pull request by carrascomj**
[eQuilibrator](https://gitlab.com/elad.noor/equilibrator-api) integration to build a _thermo database_ structure to use as `thermo_data` argument to initialize `ThermoModel`s.

* Since the thermodynamics information is directly extracted from formation energies, the uncertainty might be too high to be informative.
* No information is gathered about **cues**. eQuilibrator uses the Component Contribution method, which can't be used as in the way pytfa currently uses group contributions.
* Some tests added.

closes #36.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ branches:
 before_install:
  - pip install pytest-cov
 install:
- - pip install -e .
+ - pip install -e .[equilibrator]
```
Comment (Midnighter) [inline .travis.yml]: I recommend installing as package and not using symlinks. This let's you catch distribution errors.

```suggestion
 - pip install .[equilibrator]
```

File: pytfa/thermo/utils.py [hunk 2]
```diff
@@ -75,6 +75,7 @@ def check_reaction_balance(reaction, proton = None):
                 'Ca', 'Mn', 'Fe', 'Ni', 'Co', 'Cu', 'Zn', 'As', 'Se', 'Ag',
                 'Cd', 'W', 'Hg', 'R', 'Mo', 'X'].index(atom[0])
             except ValueError:
+                # Jorge Carrasco: Can we use warnings here?
                 print('Warning : ' + metabolite.formula + '/' + atom[0])
```
Comment (psalvy) [inline pytfa/thermo/utils.py]: Maybe use model.logger.warn instead, since we have a dedicated logger for this

File: pytfa/thermo/equilibrator.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        with keys ['deltaGf_std', 'deltaGf_err', 'error', 'struct_cues',
+        'id', 'pKa', 'mass_std', 'charge_std', 'nH_std', 'name', 'formula',
+        'other_names']
+
+    """
+    deltaGf_std, deltaGf_err = (BIGM_DG, BIGM_DG)
+    nH_std = compound.atom_bag["H"] if "H" in compound.atom_bag else 0
+    try:
+        deltaGf_std, deltaGf_err = compute_dGf(compound, cc)
+        err = "Nill"
```
Comment (psalvy) [inline pytfa/thermo/equilibrator.py]: What is this 'Nill' value for ?

File: pytfa/thermo/utils.py (hunk 2 above)
Comment (psalvy) [reply pytfa/thermo/utils.py]: `reaction.model.logger.warn` to be accurate

**Comment by psalvy** [comment]
Thank you Jorge for this. My only comment is about the warning in the reaction balance, it would be great to use the model logger instead.
I need to fix the Travis integration, and once you corrected the warning and Travis passes, we should be all clear :)
Cheers,
Pierre

**Comment by psalvy** [comment]
Hmm, actually, the PR was sent to the CI, but did not pass. Can you check this ? 
[PR build](https://travis-ci.org/EPFL-LCSB/pytfa/builds/624998879)

**Comment by psalvy** [comment]
Also, as per EPFL regulations, could you please fill in the Individual Contributor License Agreement available [here](https://github.com/EPFL-LCSB/pytfa/blob/master/CLAI)? and send it to softwares.lcsb@epfl.ch :)

File: pytfa/thermo/equilibrator.py (hunk 3 above)
Comment (carrascomj) [reply pytfa/thermo/equilibrator.py]: Oh, it is a typo. It should be "Nil", right? I took it from the [documentation](https://pytfa.readthedocs.io/en/latest/thermoDB.html#metabolites), although I am not sure if the error value is supposed be **1** if I the reaction is not to be computed.

**Comment by carrascomj** [comment]
I have just sent the CLAI file, just tell me if you need anything else.

About the CI with Codacy, I had looked into it before, but most of the _errors_ are in the `tmodel.py` and I didn't want to change it more than necessary. They are quickly to fix so I will do it now. The errors on the `equilibrator.py` are about the lazy loading, I need to take a look on Codacy, About codeclimate, it reports  "deeply nested if-else conditions" errors, but I'll see what I can do.

Furthermore, the comment about the warnings was frankly because I personally find it more convenient to work with them on notebooks than raw print statements, but the logger option is definitely better, I'll change that.

**Comment by psalvy** [comment]
Thanks for the CLAI !
No prob for the warnings, thank you for dealing with it. As for the errors, I meant the logs in Travis [here](https://travis-ci.org/EPFL-LCSB/pytfa/jobs/624998882#L413) :)
Don't worry too much about Codacy.
Cheers,
Pierre

**Comment by carrascomj** [comment]
Oh, OK, I see it now, I need to pin the version and avoid f-strings.

**Comment by psalvy** [comment]
Thank you Jorge

One last thing - rather than ignoring the test collection for Py3.5, can you instead put it as an expected fail ? In a similar manner to [this](http://doc.pytest.org/en/latest/skipping.html#skipping-test-functions) on the pytest doc.

Cheers,
Pierre

**Comment by carrascomj** [comment]
Hello,

I tried with `skipif` annotations and `importorfail` command (you can see some of my sad trial and error process on the previous commits), but it seems like the error is on the collecting phase. I can try putting the equilibrator tests in another module and manage the skipping part there, maybe that is cleaner.

**Comment by psalvy** [comment]
Good day Jorge,

Alright, we can make do with this. One last thing: Could you please supply an example using your implementation in the pytfa/tutorials folder ? That would help people get acquainted with it. 
I believe an extract from your iPython NB could be enough ?

Thank you very much again for your continued effort on this.
Cheers,
Pierre

**Comment by carrascomj** [comment]
Hmm, that didn't work out well. Maybe the python3.5 job should have a different before install definition on Travis and then skip the the equilibrator test if `equilibrator-cache` cannot be imported.

**Comment by psalvy** [comment]
Also, I was looking into why the test coverage went down from 70% to 10%. I think it is because you `import pytfa` at the beginning of your tests (see this answer [on Stackoverflow](https://stackoverflow.com/a/22288760) ) - which prevents pyTFA from being measured during the tests. Could you please try to fix that ? (importing inside the test functions instead of at the beginning of the script, as done [here](https://github.com/EPFL-LCSB/pytfa/blob/master/tests/test_io.py#L28) should be enough)

As for the Py3.5 skip, if it is too much of a hassle, we can keep your solution of skipping the collection, but make sure you specify in the install script raises an error when running on Python 3.5 :)

**Comment by psalvy** [comment]
Good day Jorge,

I hope you had a nice holiday break :) 
Any luck with respect to the test coverage ?

Cheers,
Pierre

**Comment by carrascomj** [comment]
Good day Pierre,

For now, I haven't managed to get around the problem of coverage. There are more things I can try, but the problem is that some duplications were discovered in the equilibrator compund cache, so I would prefer to wait for that to be solved before merging this PR to pytfa.

**Comment by psalvy** [comment]
No problem, thank you for your super quick update :) I just wanted to make sure not I was not the limiting step here
Cheers,
Pierre

**Comment by carrascomj** [comment]
After a huge amount of trial and error, it seems like turning pip to edit mode again works, for some reason.

File: tests/test_equilibrator.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import pytest
+import sys
+
+
+t_model = None
+
+
+# previous version aren't compatible for the required equlibrator_cache version
+# this annotation is not necessary since this file shoud be ignored in the test
+# collection phase for previous versions, but it makes it more clear.
```
Comment (psalvy) [inline tests/test_equilibrator.py]: Yes, perfect, thank you!

**Comment by psalvy** [comment]
Thank you Jorge (@carrascomj ) for this important contribution. Thanks to you, pyTFA is now compatible with another interface to obtain thermodynamic data, and you significantly contributed to make our software more accessible to others. It was a pleasure collaborating with you :)

**Pull request merged**